File: Backend/TopSys.TopConWeb.Domain/Entities/Oportunidades/Oportunidade.cs
using System;
using System.Collections.Generic;
using TopSys.TopConWeb.Domain.Entities.MotivoPerdas;
using TopSys.TopConWeb.Domain.Enums;
using TopSys.TopConWeb.Domain.Interfaces.Entities;

namespace TopSys.TopConWeb.Domain.Entities.Oportunidades
{
    public class Oportunidade : IHasEndereco
    {
        public virtual Usina Usina { get; set; }
        public int UsinaCodigo { get; set; }
        public int Numero { get; set; }
        public int Ano { get; set; }

        public int NumeroLead { get; set; }
        public int AnoLead { get; set; }

        public int NumeroVisita { get; set; }
        public int AnoVisita { get; set; }
        
        public DateTime Data { get; set; }
        public string OportunidadeNome { get; set; }
        public int IntervenienteCodigo { get; set; }
        public string Cliente { get; set; }

        public int DddTelefone { get; set; }
        public int Telefone { get; set; }
        public int DddCelular { get; set; }
        public int Celular { get; set; }
        public string Email { get; set; }
        public Vendedor Vendedor { get; set; }
        public int VendedorCodigo { get; set; }
        public Segmentacao Segmentacao { get; set; }
        public int SegmentacaoCodigo { get; set; }
        public OportunidadeTipo OportunidadeTipo { get; set; }
        public int OportunidadeTipoCodigo { get; set; }
        public CadastroGeral ViaCaptacao { get; set; }
        public int ViaCaptacaoCodigo { get; set; }
        public OportunidadeFase Fase { get; set; }
        public int FaseCodigo { get; set; }
        public EClassificacaoTemperatura Classificacao { get; set; }
        public string ProximaEtapa { get; set; }
        public DateTime? PrevisaoFechamento { get; set; }
        public MotivoPerda MotivoPerda { get; set; }
        public int? MotivoPerdaCodigo { get; set; }
        public Concorrente Concorrente { get; set; }
        public int? ConcorrenteCodigo { get; set; }
        public string ObservacaoInterna { get; set; }
        public string ObraNome { get; set; }
        public virtual CadastroGeral PorteObra { get; set; }
        public int? PorteObraCodigo { get; set; } = 0;
        public EObraFase ObraFase { get; set; }
        public float VolumeEstimadoObra { get; set; }
        public float ValorEstimadoObra { get; set; }
        public DateTime? PrevisaoInicio { get; set; }
        public DateTime? PrevisaoTermino { get; set; }
        public string EnderecoCep { get; set; }
        public string EnderecoLogradouro { get; set; }
        public int EnderecoNumero { get; set; }
        public string EnderecoComplemento { get; set; }
        public string EnderecoBairro { get; set; }
        public Municipio EnderecoMunicipio { get; set; }
        public int? EnderecoMunicipioCodigo { get; set; }
        public Usina UsinaEntrega { get; set; }
        public int? UsinaEntregaCodigo { get; set; }
        public int DistanciaUsina { get; set; }
        public string ReferenciaAcesso { get; set; }
        public OportunidadeContato ContatoPrincipal { get; set; }
        public OportunidadeContato ContatoSecundario { get; set; }

        public ICollection<OportunidadeLog> Logs { get; set; }
        public ICollection<Proposta> Propostas { get; set; }

    }
}

File: Frontend/src/app/classes/oportunidade/oportunidade.ts
import { CadastroGeral } from "../cadastro-geral/cadastro-geral";
import { Endereco } from "../endereco/endereco";
import { MotivoPerda } from "../motivo-perda/motivo-perda";
import { Proposta } from "../proposta/proposta";
import { Segmentacao } from "../segmentacao/segmentacao";
import { Usina } from "../usina/usina";
import { Vendedor } from "../vendedor/vendedor";
import { Concorrente } from "./concorrente";
import { OportunidadeContato } from "./oportunidade-contato";
import { OportunidadeFase } from "./oportunidade-fase";
import { OportunidadeTipo } from "./oportunidade-tipo";

export enum EFaseOportunidade {
    Qualificacao = 1,
    NecessitaAnalise = 2,
    Apresentacao = 3,
    Proposta = 4,
    Negociacao = 5,
    Formalizacao = 6,
    FechadoGanha = 7,
    FechadoPerdido = 8
}

export enum EClassificacaoTemperatura {
    Quente = 1,
    Morno = 2,
    Frio = 3
}

export const ClassificacaoTemperaturas = [
    { codigo: EClassificacaoTemperatura.Quente, descricao: 'QUENTE', color: '#FF8072' },
    { codigo: EClassificacaoTemperatura.Morno, descricao: 'MORNO', color: '#ffc800' },
    { codigo: EClassificacaoTemperatura.Frio, descricao: 'FRIO', color: '#9a9afa' }
];


export enum EObraFase
{
    Planejamento = 1,
    Execucao = 2,
    Conclusao = 3
}

export const FasesObra = [
    { codigo: EObraFase.Planejamento, descricao: 'PLANEJAMENTO' },
    { codigo: EObraFase.Execucao, descricao: 'EXECUCAO' },
    { codigo: EObraFase.Conclusao, descricao: 'CONCLUSAO' }
]

export class Oportunidade {
    usina: Usina;
    usinaCodigo: number = 0;

    numero: number = 0;
    ano: number = 0;

    numeroLead: number = 0;
    anoLead: number = 0;

    numeroVisita: number = 0;
    anoVisita: number = 0;

    data: Date = new Date();
    oportunidadeNome: string = '';
    intervenienteCodigo: number = 0;
    cliente: string = '';
    dddTelefone: number = 0;
    telefone: number = 0;
    dddCelular: number = 0;
    celular: number = 0;
    email: string = '';
    
    vendedor: Vendedor;
    vendedorCodigo: number = 0;

    segmentacao: Segmentacao;
    segmentacaoCodigo: number = 0;

    oportunidadeTipo: OportunidadeTipo;
    oportunidadeTipoCodigo: number = 0;

    viaCaptacao: CadastroGeral;
    viaCaptacaoCodigo: number = 0;

    fase: OportunidadeFase;
    faseCodigo: number = 0;

    classificacao: EClassificacaoTemperatura;

    proximaEtapa: string = '';
    previsaoFechamento: Date;

    motivoPerda: MotivoPerda;
    motivoPerdaCodigo: number = 0;

    concorrente: Concorrente;
    concorrenteCodigo: number = 0;

    observacaoInterna: string = '';
    obraNome: string = '';

    porteObra: CadastroGeral;
    porteObraCodigo: number = 0;

    obraFase: EObraFase;
    volumeEstimadoObra: number = 0.0;
    valorEstimadoObra: number = 0.0;
    previsaoInicio: Date;
    previsaoTermino: Date;
    endereco: Endereco = new Endereco();
    
    usinaEntrega: Usina;
    usinaEntregaCodigo: number = 0;

    propostas: Proposta[] = [];

    distanciaUsina: number = 0.0;
    referenciaAcesso: string = ''
    contatoPrincipal: OportunidadeContato = new OportunidadeContato();
    contatoSecundario: OportunidadeContato = new OportunidadeContato();
}